modules/RTC/TraceablePeerConnection.js

```diff
@@ -2568,48 +2432,77 @@ TraceablePeerConnection.prototype.setSenderVideoConstraint = function(frameHeigh
-        const ldStreamIndex = this.tpcUtils.localStreamEncodingsConfig
-            .findIndex(layer => layer.scaleResolutionDownBy === 4.0);
-
-        if (newHeight > 0 && ldStreamIndex !== -1) {
-            this.encodingsEnabledState[ldStreamIndex] = true;
-        }
+    // Set the degradation preference.
+    const preference = this.isSharingLowFpsScreen()
+        ? DEGRADATION_PREFERENCE_DESKTOP // Prefer resolution for low fps share.
+        : DEGRADATION_PREFERENCE_CAMERA; // Prefer frame-rate for high fps share and camera.
 
-        // Disable the lower spatial layers for screensharing in Unified plan when low fps screensharing is in progress
-        // There is no way to enable or disable simulcast during the call since we are re-using the same sender.
-        // Safari is an exception here since it does not send the desktop stream at all if only the high resolution
-        // stream is enabled.
-        if (this._isSharingLowFpsScreen() && this._usesUnifiedPlan && !browser.isWebKitBased()) {
-            const highResolutionEncoding = browser.isFirefox() ? 0 : this.encodingsEnabledState.length - 1;
+    parameters.degradationPreference = preference;
+    logger.info(`${this} Setting degradation preference [preference=${preference},track=${localVideoTrack}`);
 
-            this.encodingsEnabledState = this.encodingsEnabledState
-                .map((encoding, idx) => idx === highResolutionEncoding);
-        }
+    // Calculate the encodings active state based on the resolution requested by the bridge.
+    this.encodingsEnabledState = this.tpcUtils.calculateEncodingsActiveState(localVideoTrack, newHeight);
+    const presenterEnabled = localVideoTrack._originalStream
+        && localVideoTrack._originalStream.id !== localVideoTrack.getStreamId();
+    const videoType = localVideoTrack.getVideoType();
+    const desktopShareBitrate = this.options?.videoQuality?.desktopBitrate || DESKTOP_SHARE_RATE;
 
+    if (this.isSimulcastOn()) {
         for (const encoding in parameters.encodings) {
             if (parameters.encodings.hasOwnProperty(encoding)) {
                 parameters.encodings[encoding].active = this.encodingsEnabledState[encoding];
+
+                // Firefox doesn't follow the spec and lets application specify the degradation preference on the
+                // encodings.
+                browser.isFirefox() && (parameters.encodings[encoding].degradationPreference = preference);
+
+                // Max bitrates are configured on the encodings only for VP8.
+                if (this.getConfiguredVideoCodec() === CodecMimeType.VP8
+                    && (this.options?.videoQuality?.maxBitratesVideo
+                        || this.isSharingLowFpsScreen()
+                        || this._usesUnifiedPlan)) {
+                    parameters.encodings[encoding].maxBitrate = this.isSharingLowFpsScreen() && !browser.isWebKitBased()
```

Comment:
can the max bitrate calculation be extracted to a separate method just like the encoding's active states are?